File: r/index_problems.R
library("sets")

naive_loop <- function(A, B, C){

    at <- matrix(A, ncol=ncol(A), nrow=nrow(A))
    bt <- matrix(B, ncol=ncol(B), nrow=nrow(B))
    ct <- matrix(C, ncol=ncol(C), nrow=nrow(C))
    cache_scrub()
 start.time <- Sys.time()

    temp = matrix(0, nrow=dim(A)[1], ncol=dim(B)[2])
    for (i in 1:dim(C)[2]){
        temp = at %*% bt
        ct[i] = temp[i, i]
    }
    end.time <- Sys.time()
    return(tuple(as.double(end.time-start.time), ct))
}

recommended_loop <- function(A, B, C){

    at <- matrix(A, ncol=ncol(A), nrow=nrow(A))
    bt <- matrix(B, ncol=ncol(B), nrow=nrow(B))
    ct <- matrix(C, ncol=ncol(C), nrow=nrow(C))
    cache_scrub()
 start.time <- Sys.time()
    temp = at %*% bt
    for (i in 1:dim(C)[2]){
        ct[i] = temp[i, i]
    }
    end.time <- Sys.time()
    return(tuple(as.double(end.time-start.time), ct))
}

index_problems <- function(b, ...){

    res1 = benchmark("loop_nai", naive_loop, ...)
    res2 = benchmark("loop_rec", recommended_loop, ...)
    cat(c("IndexProblems correctness: ", all.equal(res1, res2), "\n"), sep="")
}

File: python/index_problems.py
from scipy import linalg
import numpy as np
import logging
import time
import copy
from benchmarker import cache_scrub

logger = logging.getLogger('IndexProblems')


def index_problems(b, *args):

    res1 = b.benchmark("loop_nai", naive_loop, *args)
    res2 = b.benchmark("loop_rec", recommended_loop, *args)
    logger.info('IndexProblems correctness: {}'.format(np.allclose(res1, res2)))


def naive_loop(A, B, C):

    at = copy.deepcopy(A)
    bt = copy.deepcopy(B)
    ct = copy.deepcopy(C)
    cache_scrub()
    start = time.perf_counter()

    temp = np.zeros((A.shape[0], B.shape[1]), dtype=np.float64)
    for i in range(C.shape[0]):
        temp = at @ bt
        ct[i] = temp[i, i]
    end = time.perf_counter()
    return end-start, ct


def recommended_loop(A, B, C):

    at = copy.deepcopy(A)
    bt = copy.deepcopy(B)
    ct = copy.deepcopy(C)
    cache_scrub()
    start = time.perf_counter()
    temp = at @ bt
    for i in range(C.shape[0]):
        ct[i] = temp[i, i]
    end = time.perf_counter()
    return end-start, ct
